File: clang/test/Refactor/ExtractRepeatedExpression/extract-repeated-expr-perform.cpp

struct AClass {
  void constMethod() const;
  void method();
};

struct AWrapper {
  const AClass &object(int x) const;
  AClass &object(int x);
};

void takesClass(AWrapper &ref) {
  int x = 0;
  ref.object(x).constMethod();
  int y = 0;
  ref.object(x).method();
}
// CHECK1: "AClass &object = ref.object(x);\nobject" [[@LINE-4]]:3 -> [[@LINE-4]]:16

// CHECK1-NEXT: "object" [[@LINE-4]]:3 -> [[@LINE-4]]:16

// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=%s:14:3 %s | FileCheck --check-prefix=CHECK1 %s
// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=%s:16:3 %s | FileCheck --check-prefix=CHECK1 %s

void variableNameSuggested(AWrapper &object) {
#ifdef IN_COMPOUND
  {
#endif
  object.object(21).constMethod();
  object.object(21).method();
#ifdef IN_COMPOUND
  }
#endif
}
// CHECK2: "AClass &object = object.object(21);\nobject" [[@LINE-6]]:3 -> [[@LINE-6]]:20

// CHECK2-NEXT: "object" [[@LINE-7]]:3 -> [[@LINE-7]]:20

// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=%s:29:3 %s | FileCheck --check-prefix=CHECK2 %s
// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=%s:29:3 %s -D IN_COMPOUND | FileCheck --check-prefix=CHECK2 %s

void takesClass2(AWrapper &ref) {
  int x = 0;
  if (x)
    ref.object(x).constMethod();
  ref.object(x).method();
}
// CHECK3: "AClass &object = ref.object(x);\n" [[@LINE-4]]:3 -> [[@LINE-4]]:3 [Symbol extracted-decl 0 1:9 -> 1:15]
// CHECK3-NEXT: "object" [[@LINE-4]]:5 -> [[@LINE-4]]:18 [Symbol extracted-decl-ref 0 1:1 -> 1:7]
// CHECK3-NEXT: "object" [[@LINE-4]]:3 -> [[@LINE-4]]:16 [Symbol extracted-decl-ref 0 1:1 -> 1:7]

// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=%s:45:5 -emit-associated %s | FileCheck --check-prefix=CHECK3 %s
// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=%s:46:3 -emit-associated %s | FileCheck --check-prefix=CHECK3 %s

void takesClass4(AWrapper &ref) {
  int x = 0;
  if (x) {
    ref.object(x).constMethod();
    ref.object(x).method();
  }
}
// CHECK4: "AClass &object = ref.object(x);\nobject" [[@LINE-4]]:5 -> [[@LINE-4]]:18

// CHECK4-NEXT: "object" [[@LINE-5]]:5 -> [[@LINE-5]]:18

// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=%s:58:5 %s | FileCheck --check-prefix=CHECK4 %s

void insertIntoCommonCompound1(AWrapper &ref) {
#ifdef EMBED
  int x = 0;
  while (true) {
#endif
  int x = 0;
  if (x) {
    if (true) {
      int y = x;
      ref.object(x).constMethod();
    }
  } else {
    ref.object(x).method();
  }
// CHECK5: "AClass &object = ref.object(x);\n" [[@LINE-8]]:3 -> [[@LINE-8]]:3
// CHECK5-NEXT: "object" [[@LINE-6]]:7 -> [[@LINE-6]]:20
// CHECK5-NEXT: "object" [[@LINE-4]]:5 -> [[@LINE-4]]:18
#ifdef EMBED
  }
#endif
}
// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=%s:77:7 %s | FileCheck --check-prefix=CHECK5 %s
// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=%s:80:5 %s -DEMBED | FileCheck --check-prefix=CHECK5 %s

void checkFirstStmtInCompoundPlacement(AWrapper &ref) {
  while (true) {
    ref.object(20);
    ref.object(20).method();
// CHECK6: "AClass &object = ref.object(20);\nobject" [[@LINE-2]]:5 -> [[@LINE-2]]:19
  }
}

// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=%s:94:5 %s | FileCheck --check-prefix=CHECK6 %s

class ImplicitThisRewrite {
  AWrapper &ref;
  ImplicitThisRewrite(AWrapper &ref) : ref(ref) {}

  void method() {
    // implicit-this: +1:5  // IMPLICIT-THIS: "AClass &object = this->ref.object(1);\nobject" [[@LINE+1]]:5 -> [[@LINE+1]]:18
    ref.object(1).method(); // IMPLICIT-NO-THIS: "AClass &object = ref.object(1);\nobject" [[@LINE]]:5 -> [[@LINE]]:18
    ref.object(1).constMethod(); // IMPLICIT-THIS-ME: "object" [[@LINE]]:5 -> [[@LINE]]:18
    // implicit-this2: +1:5
    this->ref.object(1).method(); // IMPLICIT-THIS-MENEXT: "object" [[@LINE]]:5 -> [[@LINE]]:24
  }
};

// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=implicit-this %s | FileCheck --check-prefixes=IMPLICIT-NO-THIS,IMPLICIT-THIS-ME %s
// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=implicit-this2 %s | FileCheck --check-prefixes=IMPLICIT-THIS,IMPLICIT-THIS-ME %s

File: clang/test/Refactor/ExtractRepeatedExpression/extract-repeated-expr-perform.m

@interface Object {
  int ivar;
}

- (int)instanceMethod;

@property int prop;
@property void (^block)();

@end

@interface Wrapper

- (Object *)returnsObject:(int)arg;

+ (Object *)classMethodReturnsObject;

@property(class) Object *classObject;

@property Object *object;

@end

void takesClass(Wrapper *ref) {
  int x = 0;
  [[ref returnsObject: x] instanceMethod];
  int y = x;
  [ref returnsObject: x].prop;
}
// CHECK1: "Object *duplicate = [ref returnsObject:x];\n" [[@LINE-4]]:3 -> [[@LINE-4]]:3
// CHECK1-NEXT: "duplicate" [[@LINE-5]]:4 -> [[@LINE-5]]:26
// CHECK1-NEXT: "duplicate" [[@LINE-4]]:3 -> [[@LINE-4]]:25

// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=%s:27:4 %s | FileCheck --check-prefix=CHECK1 %s
// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=%s:29:3 %s | FileCheck --check-prefix=CHECK1 %s

void takesClass2(Wrapper *ref) {
  if (2)
    [[ref object] instanceMethod];
  [ref object].block();
}
// CHECK2: "Object *object = [ref object];\n" [[@LINE-4]]:3 -> [[@LINE-4]]:3
// CHECK2-NEXT: "object" [[@LINE-4]]:6 -> [[@LINE-4]]:18
// CHECK2-NEXT: "object" [[@LINE-4]]:3 -> [[@LINE-4]]:15

// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=%s:40:6 %s | FileCheck --check-prefix=CHECK2 %s
// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=%s:41:3 %s | FileCheck --check-prefix=CHECK2 %s

void takesClass3(Wrapper *ref) {
  if (ref) {
    [ref.object instanceMethod];
    ref.object.block();
  }
}
// CHECK3: "Object *object = ref.object;\n" [[@LINE-4]]:5 -> [[@LINE-4]]:5
// CHECK3-NEXT: "object" [[@LINE-5]]:6 -> [[@LINE-5]]:16
// CHECK3-NEXT: "object" [[@LINE-5]]:5 -> [[@LINE-5]]:15

// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=%s:52:6 %s | FileCheck --check-prefix=CHECK3 %s

void worksOnClass() {
  [Wrapper classMethodReturnsObject]->ivar = 0;
  [[Wrapper classMethodReturnsObject] instanceMethod];
}
// CHECK4: "Object *classMethodReturnsObject = [Wrapper classMethodReturnsObject];\nclassMethodReturnsObject" [[@LINE-3]]:3 -> [[@LINE-3]]:37

// CHECK4-NEXT: "classMethodReturnsObject" [[@LINE-4]]:4 -> [[@LINE-4]]:38

// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=%s:63:4 %s | FileCheck --check-prefix=CHECK4 %s

void worksOnClassProperty() {
  Wrapper.classObject->ivar = 0;
  Wrapper.classObject.prop = 2;
}
// CHECK5: "Object *classObject = Wrapper.classObject;\nclassObject" [[@LINE-3]]:3 -> [[@LINE-3]]:22

// CHECK5-NEXT: "classObject" [[@LINE-4]]:3 -> [[@LINE-4]]:22

// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=%s:73:3 %s | FileCheck --check-prefix=CHECK5 %s

#define MACROARG1(X, Y) (X)
#define MACROARG2(X) X; ref.object.prop = 0

void macroArgument(Wrapper *ref) {
  // macro-arg1: +1:13
  MACROARG1(Wrapper.classObject->ivar, 1); // MACRO-ARG1: "Object *classObject = Wrapper.classObject;\n" [[@LINE]]:3 -> [[@LINE]]:3
                                             // MACRO-ARG1-NEXT: "classObject" [[@LINE-1]]:13 -> [[@LINE-1]]:32
  MACROARG1(Wrapper.classObject.prop, 0) = 0;// MACRO-ARG1-NEXT: "classObject" [[@LINE]]:13 -> [[@LINE]]:32

  // macro-arg2: +3:13
  MACROARG2(int x = 0; ref.object->ivar); // MACRO-ARG2: "Object *object = ref.object;\n" [[@LINE]]:3 -> [[@LINE]]:3
                                          // MACRO-ARG2-NEXT: "object" [[@LINE-1]]:24 -> [[@LINE-1]]:34
  MACROARG2(ref.object.prop);// MARO-ARG2-NEXT: "object" [[@LINE]]:13 -> [[@LINE]]:23
}

// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=macro-arg1 %s | FileCheck --check-prefix=MACRO-ARG1 %s
// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=macro-arg2 %s | FileCheck --check-prefix=MACRO-ARG2 %s

@interface IVarSelf {
  Wrapper *ref;
}

@end

@implementation IVarSelf

- (void)foo {
  // ivar-self: +1:3         // IVAR-SELF: "Object *object = self->ref.object;\nobject" [[@LINE+1]]:3 -> [[@LINE+1]]:13
  ref.object.prop = 0;       // IVAR-NO-SELF: "Object *object = ref.object;\nobject" [[@LINE]]:3 -> [[@LINE]]:13
  ref.object->ivar = 1;      // IVAR: "object" [[@LINE]]:3 -> [[@LINE]]:13
  // ivar-self2: +1:3
  self->ref.object.prop = 2; // IVAR-NEXT: "object" [[@LINE]]:3 -> [[@LINE]]:19
}

@end

// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=ivar-self %s | FileCheck --check-prefixes=IVAR-NO-SELF,IVAR %s
// RUN: clang-refactor-test perform -action extract-repeated-expr-into-var -at=ivar-self2 %s | FileCheck --check-prefixes=IVAR-SELF,IVAR %s
